File: java/446.arithmetic-slices-ii-subsequence.java
import java.awt.List;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Map;

/*
 * @lc app=leetcode id=446 lang=java
 *
 * [446] Arithmetic Slices II - Subsequence
 */

// @lc code=start
class Solution {
    public int numberOfArithmeticSlices(int[] nums) {
        int res = 0;
        int len = nums.length;
        if (len < 3) {
            return res;
        }
        Map<Integer, Integer>[] dp = new Map[len];
        for (int i = 0; i != len; ++i) {
            dp[i] = new HashMap<>();
            for (int j = 0; j != i; ++j) {
                var diff = (long) (nums[i]) - nums[j];
                if (diff <= Integer.MIN_VALUE || -diff <= Integer.MIN_VALUE) {
                    continue;
                }
                var intDiff = (int) diff;
                if (dp[j].containsKey(intDiff)) {
                    dp[i].put(intDiff, dp[i].getOrDefault(intDiff, 0) + dp[j].get(intDiff));
                    res += dp[j].get(intDiff);
                }
                dp[i].put(intDiff, dp[i].getOrDefault(intDiff, 0) + 1);
            }
        }
        return res;
    }
}
// @lc code=end

File: cpp/446.arithmetic-slices-ii-subsequence.cpp
/*
 * @lc app=leetcode id=446 lang=cpp
 *
 * [446] Arithmetic Slices II - Subsequence
 */
#include <vector>
#include <unordered_map>
using namespace std;
// @lc code=start
class Solution
{
public:
    int numberOfArithmeticSlices(vector<int> &nums)
    {
        int res = 0;
        int len = nums.size();
        if (len < 3)
        {
            return res;
        }
        vector<unordered_map<int, int>> dp(len);

        for (int i = 0; i != len; ++i)
        {
            for (int j = 0; j != i; ++j)
            {
                auto diff = static_cast<long>(nums[i]) - nums[j];
                if (diff <= INT32_MIN || -diff <= INT32_MIN)
                {
                    continue;
                }
                if (dp[j].count(diff))
                {
                    dp[i][diff] += dp[j][diff];
                    res += dp[j][diff];
                }
                dp[i][diff] += 1;
            }
        }
        return res;
    }
};
// @lc code=end
